README.md
[![Build Status](https://travis-ci.com/pulp-platform/common_cells.svg?branch=master)](https://travis-ci.com/pulp-platform/common_cells)

# Common Cells Repository

Maintainer: Florian Zaruba <zarubaf@iis.ee.ethz.ch>

This repository contains commonly used cells and headers for use in various projects.

## Cell Contents

This repository currently contains the following cells, ordered by categories.
Please note that cells with status *deprecated* are not to be used for new designs and only serve to provide compatibility with old code.

### Clocks and Resets

|           Name          |                     Description                     |    Status    | Superseded By |
|-------------------------|-----------------------------------------------------|--------------|---------------|
| `clk_div`               | Clock divider with integer divisor                  | active       |               |
| `clock_divider`         | Clock divider with configuration registers          | *deprecated* | `clk_div`     |
| `clock_divider_counter` | Clock divider using a counter                       | *deprecated* | `clk_div`     |
| `rstgen`                | Reset synchronizer                                  | active       |               |
| `rstgen_bypass`         | Reset synchronizer with dedicated test reset bypass | active       |               |

### Clock Domains and Asynchronous Crossings

|         Name         |                                   Description                                    |    Status    | Superseded By |
|----------------------|----------------------------------------------------------------------------------|--------------|---------------|
| `cdc_2phase`         | Clock domain crossing using two-phase handshake, with ready/valid interface      | active       |               |
| `cdc_fifo_2phase`    | Clock domain crossing FIFO using two-phase handshake, with ready/valid interface | active       |               |
| `cdc_fifo_gray`      | Clock domain crossing FIFO using a gray-counter, with ready/valid interface      | active       |               |
| `edge_detect`        | Rising/falling edge detector                                                     | active       |               |
| `edge_propagator`    | **ANTONIO ADD DESCRIPTION**                                                      | active       |               |
| `edge_propagator_rx` | **ANTONIO ADD DESCRIPTION**                                                      | active       |               |
| `edge_propagator_tx` | **ANTONIO ADD DESCRIPTION**                                                      | active       |               |
| `pulp_sync`          | Serial line synchronizer                                                         | *deprecated* | `sync`        |
| `pulp_sync_wedge`    | Serial line synchronizer with edge detector                                      | *deprecated* | `sync_wedge`  |
| `serial_deglitch`    | Serial line deglitcher                                                           | active       |               |
| `sync`               | Serial line synchronizer                                                         | active       |               |
| `sync_wedge`         | Serial line synchronizer with edge detector                                      | active       |               |

### Counters and Shift Registers

|         Name        |                   Description                                     |    Status    | Superseded By |
|---------------------|-------------------------------------------------------------------|--------------|---------------|
| `counter`           | Generic up/down counter with overflow detection                   | active       |               |
| `delta_counter`     | Up/down counter with variable delta and overflow detection        | active       |               |
| `generic_LFSR_8bit` | 8-bit linear feedback shift register (LFSR)                       | *deprecated* | `lfsr_8bit`   |
| `lfsr_8bit`         | 8-bit linear feedback shift register (LFSR)                       | active       |               |
| `lfsr_16bit`        | 16-bit linear feedback shift register (LFSR)                      | active       |               |
| `lfsr`              | 4...64-bit parametric Galois LFSR with optional whitening feature | active       |               |
| `max_counter`       | Up/down counter with variable delta that tracks its maximum value | active       |               |
| `mv_filter`         | **ZARUBAF ADD DESCRIPTION**                                       | active       |               |

### Data Path Elements

| Name                         | Description                                                                    | Status         | Superseded By |
| :--------------------------- | :----------------------------------------------------------------------------- | :------------- | :------------ |
| `addr_decode   `             | Address map decoder                                                            | active         |               |
| `ecc_decode`                 | SECDED Decoder (Single Error Correction, Double Error Detection)               | active         |               |
| `ecc_encode`                 | SECDED Encoder (Single Error Correction, Double Error Detection)               | active         |               |
| `binary_to_gray`             | Binary to gray code converter                                                  | active         |               |
| `find_first_one`             | Leading-one finder / leading-zero counter                                      | *deprecated*   | `lzc`         |
| `gray_to_binary`             | Gray code to binary converter                                                  | active         |               |
| `lzc`                        | Leading/trailing-zero counter                                                  | active         |               |
| `onehot_to_bin`              | One-hot to binary converter                                                    | active         |               |
| `shift_reg`                  | Shift register for arbitrary types                                             | active         |               |
| `rr_arb_tree`                | Round-robin arbiter for req/gnt and vld/rdy interfaces with optional priority  | active         |               |
| `rrarbiter`                  | Round-robin arbiter for req/ack interface with look-ahead                      | *deprecated*   | `rr_arb_tree` |
| `prioarbiter`                | Priority arbiter arbiter for req/ack interface with look-ahead                 | *deprecated*   | `rr_arb_tree` |
| `fall_through_register`      | Fall-through register with ready/valid interface                               | active         |               |
| `spill_register`             | Register with ready/valid interface to cut all combinational interface paths   | active         |               |
| `stream_arbiter`             | Round-robin arbiter for ready/valid stream interface                           | active         |               |
| `stream_arbiter_flushable`   | Round-robin arbiter for ready/valid stream interface and flush functionality   | active         |               |
| `stream_demux`               | Ready/valid interface demultiplexer                                            | active         |               |
| `stream_mux`                 | Ready/valid interface multiplexer                                              | active         |               |
| `stream_register`            | Register with ready/valid interface                                            | active         |               |
| `stream_fork`                | Ready/valid fork                                                               | active         |               |
| `stream_filter`              | Ready/valid filter                                                             | active         |               |
| `stream_delay`               | Randomize or delay ready/valid interface                                       | active         |               |
| `sub_per_hash`               | Substitution-permutation hash function                                         | active         |               |
| `popcount`                   | Combinatorial popcount (hamming weight)                                        | active         |               |

### Data Structures

| Name                 | Description                                     | Status         | Superseded By |
| :------------------- | :---------------------------------------------- | :------------- | :------------ |
| `cb_filter`          | Counting-Bloom-Filter with combinational lookup | active         |               |
| `fifo`               | FIFO register with upper threshold              | *deprecated*   | `fifo_v3`     |
| `fifo_v2`            | FIFO register with upper and lower threshold    | *deprecated*   | `fifo_v3`     |
| `fifo_v3`            | FIFO register with generic fill counts          | active         |               |
| `generic_fifo`       | FIFO register without thresholds                | *deprecated*   | `fifo_v3`     |
| `generic_fifo_adv`   | FIFO register without thresholds                | *deprecated*   | `fifo_v3`     |
| `sram`               | SRAM behavioral model                           | active         |               |
| `plru_tree`          | Pseudo least recently used tree                 | active         |               |
| `unread`             | Empty module to sink unconnected outputs into   | active         |               |


## Header Contents

This repository currently contains the following header files.

### RTL Register Macros

The header file `registers.svh` contains macros that expand to descriptions of registers.
To avoid misuse of `always_ff` blocks, only the following macros shall be used to describe sequential behavior.
The use of linter rules that flag explicit uses of `always_ff` in source code is encouraged.

|         Macro         |                            Arguments                            |                               Description                               |
|-----------------------|-----------------------------------------------------------------|-------------------------------------------------------------------------|
| <code>\`FF</code>     | `q_sig`, `d_sig`, `rst_val`                                     | Flip-flop with asynchronous active-low reset (implicit)                 |
| <code>\`FFAR</code>   | `q_sig`, `d_sig`, `rst_val`, `clk_sig`, `arst_sig`              | Flip-flop with asynchronous active-high reset                           |
| <code>\`FFARN</code>  | `q_sig`, `d_sig`, `rst_val`, `clk_sig`, `arstn_sig`             | Flip-flop with asynchronous active-low reset                            |
| <code>\`FFSR</code>   | `q_sig`, `d_sig`, `rst_val`, `clk_sig`, `rst_sig`               | Flip-flop with synchronous active-high reset                            |
| <code>\`FFSRN</code>  | `q_sig`, `d_sig`, `rst_val`, `clk_sig`, `rstn_sig`              | Flip-flop with synchronous active-low reset                             |
| <code>\`FFNR</code>   | `q_sig`, `d_sig`, `clk_sig`                                     | Flip-flop without reset                                                 |
|                       |                                                                 |                                                                         |
| <code>\`FFL</code>    | `q_sig`, `d_sig`, `load_ena`, `rst_val`                         | Flip-flop with load-enable and asynchronous active-low reset (implicit) |
| <code>\`FFLAR</code>  | `q_sig`, `d_sig`, `load_ena`, `rst_val`, `clk_sig`, `arst_sig`  | Flip-flop with load-enable and asynchronous active-high reset           |
| <code>\`FFLARN</code> | `q_sig`, `d_sig`, `load_ena`, `rst_val`, `clk_sig`, `arstn_sig` | Flip-flop with load-enable and asynchronous active-low reset            |
| <code>\`FFLSR</code>  | `q_sig`, `d_sig`, `load_ena`, `rst_val`, `clk_sig`, `rst_sig`   | Flip-flop with load-enable and synchronous active-high reset            |
| <code>\`FFLSRN</code> | `q_sig`, `d_sig`, `load_ena`, `rst_val`, `clk_sig`, `rstn_sig`  | Flip-flop with load-enable and synchronous active-low reset             |
| <code>\`FFLNR</code>  | `q_sig`, `d_sig`, `load_ena`, `clk_sig`                         | Flip-flop with load-enable without reset                                |
- *The name of the clock and reset signals for implicit variants is `clk_i` and `rst_ni`, respectively.*
- *Argument suffix `_sig` indicates signal names for present and next state as well as clocks and resets.*
- *Argument `rst_val` specifies the value literal to be assigned upon reset.*
- *Argument `load_ena` specifies the boolean expression that forms the load enable of the register.*


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## Unreleased
### Added
- SECDED (Single Error Correction, Double Error Detection) encoder and decoder
- SECDED Verilator-based testbench
- Travis build for SECDED module

### Fixed

## 1.16.4 - 2020-03-02
### Fixed
- id_queue: Fix generation of `head_tail_q` registers

## 1.16.3 - 2020-02-11
### Fixed
- Handle degenerated `addr_decode` with `NoIndices == 1`, change default parameters to `32'd0`

## 1.16.2 - 2020-02-04
### Fixed
- Fix author section in Bender.yml

## 1.16.1 - 2020-02-03
### Fixed
- `rr_arb_tree`: Add guard SVA statement for Verilator
- Added missing sources in `Bender.yml` and `src_files.yml`

## 1.16.0 - 2020-01-13
### Fixed
- Handle degenerated `onehot_to_bin` with `ONEHOT_WIDTH == 1`
- Handle degenerated `id_queue` with `CAPACITY == 1` or `HT_CAPACITY == 1`
- Fix `cdc_fifo_gray` to be a safe clock domain crossing (CDC)

## 1.15.0 - 2019-12-09
### Added
- Added address map decoder module

### Fixed
- Handle degenerated `lzc` with `WIDTH == 1`

## 1.14.0 - 2019-10-08

### Added
- Added spubstitution-permutation hash function module
- Added couning-bloom-filter module
- `spill_register`: Added Bypass parameter
- `counter`: Added sticky overflow
- Added counter with variable delta
- Added counter that tracks its maximum value

### Changed
- Added formal testbench for `fifo` and `fall_through_regsiter`

## 1.13.1 - 2019-06-01

### Changed

- Fix path in `src_files.yml` for `stream_arbiter` and `stream_arbiter_flushable`

## 1.13.0 - 2019-05-29

### Added

- Added exponential backoff window module
- Added parametric Galois LFSR module with optional whitening feature
- Added `cf_math_pkg`: Constant Function implementations of mathematical functions for HDL elaboration

### Changed
- Parametric payload data type for `rr_arb_tree`

### Deprecated
- The following arbiter implementations are deprecated and superseded by `rr_arb_tree`:
- Priority arbiter `prioarbiter`
- Round-robin arbiter `rrarbiter`

### Fixed

## 1.12.0 - 2019-04-09

### Added
- Add priority arbiter
- Add Pseudo Least Recently Used tree
- Add round robin arbiter mux tree

### Changed
- Add selectable arbiter implementation for `stream_arbiter` and `stream_arbiter_flushable`. One can choose between priority (`prio`) and round-robin arbitration (`rr`).
- Add `$onehot0` assertion in one-hot to bin
- Rework `rrarbiter` unit (uses `rr_arb_tree` implementation underneath)

## 1.11.0 - 2019-03-20

### Added
- Add stream fork
- Add fall-through register
- Add stream filter
- Add ID queue

### Changed
- `sync_wedge` use existing synchronizer. This defines a single place where a tech-specific synchronizer can be defined.

### Fixed
- Fix FIFO push and pop signals in `stream_register` to observe interface prerequisites.
- In `fifo_v3`, fix data output when pushing into empty fall-through FIFO. Previously, the data
  output of an empty fall-through FIFO with data at its input (and `push_i=1`) depended on
  `pop_i`: When `pop_i=0`, old, invalid data were visible at the output (even though `empty_o=0`,
  indicating that the data output is valid). Only when `pop_i=1`, the data from the input fell
  through. One consequence of this bug was that `data_o` of the `fall_through_register` could change
  while `valid_o=1`, violating the basic stream specification.

## 1.10.0 - 2018-12-18

### Added
- Add `fifo_v3` with generic fill count
- Add 16 bit LFSR
- Add stream delayer
- Add stream arbiter
- Add register macros for RTL
- Add shift register

### Changed
- Make number of registers of `rstgen_bypass` a parameter.

### Fixed
- Fix `valid_i` and `grant_i` guarantees in `generic_fifo` for backward compatibility.
- LZC: Synthesis of streaming operators in ternary operators
- Add missing entry for `popcount` to `Bender.yml`.
- Add default values for parameters to improve compatibility with Synopsys DC and Vivado.

## 1.9.0 - 2018-11-02

### Added
- Add popcount circuit `popcount`

## 1.8.0 - 2018-10-15

### Added
- Add lock feature to the rrarbiter. This prevents the arbiter to change the decision when we have pending requests that remain unaknowledged for several cycles.
- Add deglitching circuit
- Add generic clock divider
- Add edge detecter as alias to sync_wedge (name is more expressive)
- Add generic counter
- Add moving deglitcher

## 1.7.6 - 2018-09-27

### Added
- Add reset synchronizer with explicit reset bypass in testmode

## 1.7.5 - 2018-09-06
### Fixed
- Fix incompatibility with verilator
- Fix dependency to open-source repo

## 1.7.4 - 2018-09-06
- Fix assertions in `fifo_v2` (write on full / read on empty did not trigger properly)

## 1.7.3 - 2018-08-27
### Fixed
- Use proper `fifo_v2` in `generic_fifo` module.

## 1.7.2 - 2018-08-27
### Added
- Almost full/empty flags to FIFO, as `fifo_v2`.

### Changed
- FIFO moved to `fifo_v1` and instantiates `fifo_v2`.

## 1.7.1 - 2018-08-27
### Fixed
- Revert breaking changes to `fifo`.

## 1.7.0 - 2018-08-24
### Added
- Add stream register (`stream_register`).
- Add stream multiplexer and demultiplexer (`stream_mux`, `stream_demux`).
- Add round robin arbiter (`rrarbiter`).
- Add leading zero counter (`lzc`).

### Changed
- Deprecate `find_first_one` in favor of `lzc`.

## 1.6.0 - 2018-04-03
### Added
- Add binary to Gray code converter.
- Add Gray code to binary converter.
- Add Gray code testbench.
- Add CDC FIFO based on Gray counters. This is a faster alternative to the 2-phase FIFO which also works if a domain's clock has stopped.

### Changed
- Rename `cdc_fifo` to `cdc_fifo_2phase`.
- Adjust CDC FIFO testbench to cover both implementations.

## 1.5.4 - 2018-03-31
### Changed
- Replace explicit clock gate in `fifo` with implicit one.

## 1.5.3 - 2018-03-16
### Changed
- Remove duplicate deprecated modules.


test/ecc/ecc.h
#include <stdint.h>
#include <stdlib.h>

bool ispowerof2(unsigned int x);
void toBin(uint8_t *data, uint64_t word, size_t len);
void toByte(uint8_t *data, uint8_t *in_data, size_t len);
void printBits(const char *preamble, uint8_t *data, size_t len);
int calcParityLen(int data_bits);
void ecc_encode (uint64_t word, int d, uint8_t *codeword);

test/synth.tcl
remove_design -all
set link_library []
# This script was generated automatically by bender.
set search_path_initial $search_path
set ROOT "/home/zarubaf/common_cells"

set search_path $search_path_initial
lappend search_path "$ROOT/include"

analyze -format sv \
    -define { \
        TARGET_SYNOPSYS \
        TARGET_SYNTHESIS \
        TARGET_SYNTH_TEST \
    } \
    [list \
        "$ROOT/.bender/git/checkouts/tech_cells_generic-0bad0132c3cc22f2/src/deprecated/cluster_clk_cells.sv" \
        "$ROOT/.bender/git/checkouts/tech_cells_generic-0bad0132c3cc22f2/src/deprecated/pulp_clk_cells.sv" \
        "$ROOT/.bender/git/checkouts/tech_cells_generic-0bad0132c3cc22f2/src/rtl/tc_clk.sv" \
    ]

set search_path $search_path_initial
lappend search_path "$ROOT/include"

analyze -format sv \
    -define { \
        TARGET_SYNOPSYS \
        TARGET_SYNTHESIS \
        TARGET_SYNTH_TEST \
    } \
    [list \
        "$ROOT/src/addr_decode.sv" \
        "$ROOT/src/cdc_2phase.sv" \
        "$ROOT/src/cf_math_pkg.sv" \
        "$ROOT/src/clk_div.sv" \
        "$ROOT/src/delta_counter.sv" \
        "$ROOT/src/ecc_pkg.sv" \
        "$ROOT/src/edge_propagator_tx.sv" \
        "$ROOT/src/exp_backoff.sv" \
        "$ROOT/src/fifo_v3.sv" \
        "$ROOT/src/graycode.sv" \
        "$ROOT/src/lfsr.sv" \
        "$ROOT/src/lfsr_16bit.sv" \
        "$ROOT/src/lfsr_8bit.sv" \
        "$ROOT/src/lzc.sv" \
        "$ROOT/src/mv_filter.sv" \
        "$ROOT/src/onehot_to_bin.sv" \
        "$ROOT/src/plru_tree.sv" \
        "$ROOT/src/popcount.sv" \
        "$ROOT/src/rr_arb_tree.sv" \
        "$ROOT/src/rstgen_bypass.sv" \
        "$ROOT/src/serial_deglitch.sv" \
        "$ROOT/src/shift_reg.sv" \
        "$ROOT/src/spill_register.sv" \
        "$ROOT/src/stream_demux.sv" \
        "$ROOT/src/stream_filter.sv" \
        "$ROOT/src/stream_fork.sv" \
        "$ROOT/src/stream_mux.sv" \
        "$ROOT/src/sub_per_hash.sv" \
        "$ROOT/src/sync.sv" \
        "$ROOT/src/sync_wedge.sv" \
        "$ROOT/src/unread.sv" \
        "$ROOT/src/cb_filter.sv" \
        "$ROOT/src/cdc_fifo_2phase.sv" \
        "$ROOT/src/cdc_fifo_gray.sv" \
        "$ROOT/src/counter.sv" \
        "$ROOT/src/ecc_decode.sv" \
        "$ROOT/src/ecc_encode.sv" \
        "$ROOT/src/edge_detect.sv" \
        "$ROOT/src/id_queue.sv" \
        "$ROOT/src/max_counter.sv" \
        "$ROOT/src/rstgen.sv" \
        "$ROOT/src/stream_delay.sv" \
        "$ROOT/src/fall_through_register.sv" \
        "$ROOT/src/stream_arbiter_flushable.sv" \
        "$ROOT/src/stream_register.sv" \
        "$ROOT/src/stream_arbiter.sv" \
    ]

set search_path $search_path_initial
lappend search_path "$ROOT/include"

analyze -format sv \
    -define { \
        TARGET_SYNOPSYS \
        TARGET_SYNTHESIS \
        TARGET_SYNTH_TEST \
    } \
    [list \
        "$ROOT/test/cdc_2phase_synth.sv" \
        "$ROOT/test/id_queue_synth.sv" \
        "$ROOT/test/stream_arbiter_synth.sv" \
        "$ROOT/test/ecc_synth.sv" \
        "$ROOT/test/synth_bench.sv" \
    ]

set search_path $search_path_initial
lappend search_path "$ROOT/include"

analyze -format sv \
    -define { \
        TARGET_SYNOPSYS \
        TARGET_SYNTHESIS \
        TARGET_SYNTH_TEST \
    } \
    [list \
        "$ROOT/src/deprecated/clock_divider_counter.sv" \
        "$ROOT/src/deprecated/find_first_one.sv" \
        "$ROOT/src/deprecated/generic_LFSR_8bit.sv" \
        "$ROOT/src/deprecated/generic_fifo.sv" \
        "$ROOT/src/deprecated/prioarbiter.sv" \
        "$ROOT/src/deprecated/pulp_sync.sv" \
        "$ROOT/src/deprecated/pulp_sync_wedge.sv" \
        "$ROOT/src/deprecated/rrarbiter.sv" \
        "$ROOT/src/deprecated/clock_divider.sv" \
        "$ROOT/src/deprecated/fifo_v2.sv" \
        "$ROOT/src/deprecated/fifo_v1.sv" \
        "$ROOT/src/edge_propagator.sv" \
        "$ROOT/src/edge_propagator_rx.sv" \
    ]

set search_path $search_path_initial
elaborate synth_bench


test/ecc_synth.sv
// Copyright 2020 ETH Zurich and University of Bologna.
// Copyright and related rights are licensed under the Solderpad Hardware
// License, Version 0.51 (the "License"); you may not use this file except in
// compliance with the License. You may obtain a copy of the License at
// http://solderpad.org/licenses/SHL-0.51. Unless required by applicable law
// or agreed to in writing, software, hardware and materials distributed under
// this License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
// CONDITIONS OF ANY KIND, either express or implied. See the License for the
// specific language governing permissions and limitations under the License.

module ecc_synth;

    for (genvar dw = 8; dw < 32; dw += 8) begin
        logic [ecc_pkg::get_cw_width(dw):0] codeword;
        ecc_encode #(.DataWidth(dw)) i_encode (.data_i(), .data_o(codeword));
        ecc_decode #(.DataWidth(dw))
        i_decode (
            .data_i (codeword),
            .data_o (),
            .syndrome_o (),
            .single_error_o (),
            .double_error_o ()
        );
    end

endmodule

src/ecc_pkg.sv
// Copyright 2020 ETH Zurich and University of Bologna.
// Copyright and related rights are licensed under the Solderpad Hardware
// License, Version 0.51 (the "License"); you may not use this file except in
// compliance with the License. You may obtain a copy of the License at
// http://solderpad.org/licenses/SHL-0.51. Unless required by applicable law
// or agreed to in writing, software, hardware and materials distributed under
// this License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
// CONDITIONS OF ANY KIND, either express or implied. See the License for the
// specific language governing permissions and limitations under the License.
//
// Author: Florian Zaruba <zarubaf@iis.ee.ethz.ch>
//
// Description: Contains common ECC definitions and helper functions.

package ecc_pkg;

  // Calculate required ECC parity width:
  function automatic int unsigned get_parity_width (input int unsigned data_width);
    // data_width + cw_width + 1 <= 2**cw_width
    int unsigned cw_width = 2;
    while (unsigned'(2**cw_width) < cw_width + data_width + 1) cw_width++;
    return cw_width;
  endfunction

  // Calculate required ECC codeword width:
  function automatic int unsigned get_cw_width (input int unsigned data_width);
    // data width + parity width + one additional parity bit (for double error detection)
    return data_width + get_parity_width(data_width) + 1;
  endfunction

endpackage


src/ecc_encode.sv
// Copyright 2020 ETH Zurich and University of Bologna.
// Copyright and related rights are licensed under the Solderpad Hardware
// License, Version 0.51 (the "License"); you may not use this file except in
// compliance with the License. You may obtain a copy of the License at
// http://solderpad.org/licenses/SHL-0.51. Unless required by applicable law
// or agreed to in writing, software, hardware and materials distributed under
// this License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
// CONDITIONS OF ANY KIND, either express or implied. See the License for the
// specific language governing permissions and limitations under the License.
//
// Author: Florian Zaruba <zarubaf@iis.ee.ethz.ch>
// ECC Encoder
//
// Implements SECDED (Single Error Correction, Double Error Detection) Hamming Code
// with extended parity bit [1].
// The module receives a data word and encodes it using above mentioned error
// detection and correction code. The corresponding decode module
// can be found in `ecc_decode.sv`
//
// [1] https://en.wikipedia.org/wiki/Hamming_code

module ecc_encode import ecc_pkg::*; #(
  parameter  int unsigned DataWidth   = 64,
  // Do not change
  parameter type data_t         = logic [DataWidth-1:0],
  parameter type parity_t       = logic [get_parity_width(DataWidth)-1:0],
  parameter type code_word_t    = logic [get_cw_width(DataWidth)-2:0],
  parameter type encoded_data_t = struct packed {
                                    logic parity;
                                    code_word_t code_word;
                                  }
) (
  input  data_t         data_i,
  output encoded_data_t data_o
);

  parity_t parity_code_word;
  code_word_t data, codeword;

  // Expand incoming data to codeword width
  always_comb begin : expand_data
    automatic int unsigned idx;
    data = '0;
    idx = 0;
    for (int unsigned i = 1; i < unsigned'($bits(code_word_t)) + 1; i++) begin
      // if it is not a power of two word it is a normal data index
      if (unsigned'(2**$clog2(i)) != i) begin
        data[i - 1] = data_i[idx];
        idx++;
      end
    end
  end

  // calculate code word
  always_comb begin : calculate_syndrome
    parity_code_word = 0;
    for (int unsigned i = 0; i < unsigned'($bits(parity_t)); i++) begin
      for (int unsigned j = 1; j < unsigned'($bits(code_word_t)) + 1; j++) begin
        if (|(unsigned'(2**i) & j)) parity_code_word[i] = parity_code_word[i] ^ data[j - 1];
      end
    end
  end

  // fuse the final codeword
  always_comb begin : generate_codeword
      codeword = data;
      for (int unsigned i = 0; i < unsigned'($bits(parity_t)); i++) begin
        codeword[2**i-1] = parity_code_word[i];
      end
  end

  assign data_o.code_word = codeword;
  assign data_o.parity = ^codeword;

endmodule


src/ecc_decode.sv
// Copyright 2020 ETH Zurich and University of Bologna.
// Copyright and related rights are licensed under the Solderpad Hardware
// License, Version 0.51 (the "License"); you may not use this file except in
// compliance with the License. You may obtain a copy of the License at
// http://solderpad.org/licenses/SHL-0.51. Unless required by applicable law
// or agreed to in writing, software, hardware and materials distributed under
// this License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
// CONDITIONS OF ANY KIND, either express or implied. See the License for the
// specific language governing permissions and limitations under the License.
//
// Author: Florian Zaruba <zarubaf@iis.ee.ethz.ch>
//
// ECC Decoder
//
// Implements SECDED (Single Error Correction, Double Error Detection) Hamming Code
// with extended parity bit [1].
// The module receives a data word including parity bit and decodes it according to the
// number of data and parity bit.
//
// 1. If no error has been detected the syndrome will be zero and all flags will be zero.
// 2. If a single error has been detected, the syndrome is non-zero, and `single-error` will be
//    asserted. The output word contains the corrected data.
// 3. If the parity bit contained an error, the module will assert `parity_error_o`.
// 4. In case of a double fault the syndrome is non-zero, `double_error_o` will be asserted.
//    All other status flags will be de-asserted.
//
// [1] https://en.wikipedia.org/wiki/Hamming_code

module ecc_decode import ecc_pkg::*; #(
  parameter  int unsigned DataWidth   = 64,
  // Do not change
  parameter type data_t         = logic [DataWidth-1:0],
  parameter type parity_t       = logic [get_parity_width(DataWidth)-1:0],
  parameter type code_word_t    = logic [get_cw_width(DataWidth)-2:0],
  parameter type encoded_data_t = struct packed {
                                    logic parity;
                                    code_word_t code_word;
                                  }
  ) (
  input  encoded_data_t data_i,
  output data_t         data_o,
  output parity_t       syndrome_o, // indicates the erroneous bit position
  output logic          single_error_o,
  output logic          parity_error_o, // error received in parity bit (MSB)
  output logic          double_error_o
);

  logic       parity;
  data_t      data_wo_parity;
  parity_t    syndrome;
  logic       syndrome_not_zero;
  code_word_t correct_data;

  // Check parity bit. 0 = parity equal, 1 = different parity
  assign parity = data_i.parity ^ (^data_i.code_word);

  //    | 0  1  2  3  4  5  6  7  8  9 10 11 12  13  14
  //    |p1 p2 d1 p4 d2 d3 d4 p8 d5 d6 d7 d8 d9 d10 d11
  // ---|----------------------------------------------
  // p1 | x     x     x     x     x     x     x       x
  // p2 |    x  x        x  x        x  x         x   x
  // p4 |          x  x  x  x              x  x   x   x
  // p8 |                      x  x  x  x  x  x   x   x

  // 1. Parity bit 1 covers all bit positions which have the least significant bit
  //    set: bit 1 (the parity bit itself), 3, 5, 7, 9, etc.
  // 2. Parity bit 2 covers all bit positions which have the second least
  //    significant bit set: bit 2 (the parity bit itself), 3, 6, 7, 10, 11, etc.
  // 3. Parity bit 4 covers all bit positions which have the third least
  //    significant bit set: bits 4–7, 12–15, 20–23, etc.
  // 4. Parity bit 8 covers all bit positions which have the fourth least
  //    significant bit set: bits 8–15, 24–31, 40–47, etc.
  // 5. In general each parity bit covers all bits where the bitwise AND of the
  //    parity position and the bit position is non-zero.
  always_comb begin : calculate_syndrome
    syndrome = 0;
    for (int unsigned i = 0; i < unsigned'($bits(parity_t)); i++) begin
      for (int unsigned j = 0; j < unsigned'($bits(code_word_t)); j++) begin
        if (|(unsigned'(2**i) & (j + 1))) syndrome[i] = syndrome[i] ^ data_i.code_word[j];
      end
    end
  end

  assign syndrome_not_zero = |syndrome;

  // correct the data word if the syndrome is non-zero
  always_comb begin
    correct_data = data_i.code_word;
    if (syndrome_not_zero) begin
      correct_data[syndrome - 1] = ~data_i.code_word[syndrome - 1];
    end
  end

  // Syndrome | Overall Parity (MSB) | Error Type   | Notes
  // --------------------------------------------------------
  // 0        | 0                    | No Error     |
  // /=0      | 1                    | Single Error | Correctable. Syndrome holds incorrect bit position.
  // 0        | 1                    | Parity Error | Overall parity, MSB is in error and can be corrected.
  // /=0      | 0                    | Double Error | Not correctable.
  assign single_error_o = parity & syndrome_not_zero;
  assign parity_error_o = parity & ~syndrome_not_zero;
  assign double_error_o = ~parity & syndrome_not_zero;

  // Extract data vector
  always_comb begin
    automatic int unsigned idx; // bit index
    data_wo_parity = '0;
    idx = 0;

    for (int unsigned i = 1; i < unsigned'($bits(code_word_t)) + 1; i++) begin
      // if i is a power of two we are indexing a parity bit
      if (unsigned'(2**$clog2(i)) != i) begin
        data_wo_parity[idx] = correct_data[i - 1];
        idx++;
      end
    end
  end

  assign data_o = data_wo_parity;

endmodule
